hera_sim/interpolators.py

```diff
@@ -0,0 +1,99 @@
+"""
+This module provides interfaces to different interpolation classes.
+"""
+
+import numpy as np
+from scipy.interpolate import RectBivariateSpline
+
+class Tsky:
+    """
+    # TODO: fill in docstring
+    """
+    def __init__(self, datafile, **interp_kwargs):
+        self.datafile = datafile
+        self._data = np.load(self.datafile, allow_pickle=True)
+        # make sure the correct keys are used
+        assert 'freqs' in self._data.keys(), \
+                "The frequencies corresponding to the sky temperature array " \
+                "must be provided. They must be saved to the npz file using " \
+                "the key 'freqs'."
+        assert 'lsts' in self._data.keys(), \
+                "The LSTs corresponding to the sky temperature array must " \
+                "be provided. They must be saved to the npz file using the " \
+                "key 'lsts'."
+        assert 'tsky' in self._data.keys(), \
+                "The sky temperature array must be saved to the npz file " \
+                "using the key 'tsky'."
+        assert 'meta' in self._data.keys(), \
+                "The npz file must contain a metadata dictionary that can " \
+                "be accessed with the key 'meta'. This dictionary should " \
+                "provide information about the units of the various arrays " \
+                "and the polarizations of the sky temperature array."
+        # now make sure that the sky temperature array has the correct shape
+        # this should be (NPOLS, NLSTS, NFREQS)
+        # XXX this is somewhat strict, but useful in the case NFREQS==NLSTS
+        # XXX however, in the case NFREQS==NLSTS, it is possible for the
+        # XXX sky temperature to be shaped incorrectly and pass this assertion
+        assert self.tsky.shape==(len(self.meta['pols']),
+                                 self.lsts.size, self.freqs.size), \
+                "The tsky array is incorrectly shaped. Please ensure that " \
+                "the tsky array has shape (NPOLS, NLSTS, NFREQS)."
+
+        self._interpolators = {}
+
+        # fill in interpolators
+        # XXX do we want to make a hidden member function for generating
+        # XXX the interpolator objects?
+        for j, pol in enumerate(self.meta['pols']):
+            tsky_data = self.tsky[j]
+
+            # fill in information outside of measured LST values for
+            # interpolator stability
+            lsts = np.concatenate([self.lsts[-10:]-2*np.pi,
+                                   self.lsts,
+                                   self.lsts[:10]+2*np.pi])
+            tsky_data = np.concatenate([tsky_data[-10:],
+                                        tsky_data,
+                                        tsky_data[:10]])
+            
+            # now make the interpolation object
+            self._interpolators[pol] = RectBivariateSpline(lsts, 
```

Comment:
Also, creating the spline object can take some time. If the user never needs a certain pol, then doing this for every pol is unnecessary. There's two ways around this: the first idea is to pass a non-changeable 'pol' keyword to `__init__`, which would set which polarization this particular instance uses once and for all. This is the easiest way, and is more easily specifiable via YAML. However, it does lack some flexibility, in that once the object is created you can't get at the other polarization.

The other idea would be to make a method:
```
@memoize
def _interpolator(self, pol='xx'):
    j = self.meta['pols'].index(pol)
    < do all the stuff you did to get a single interpolator >
```

Then everything else as you've got it, except to replace calls to `self._interpolators['xx']` with `self._interpolator('xx')`. The `@memoize` function caches results based on the inputs. This means that if 'yy' is never used, it is never calculated, but the 'xx' is only ever calculated once.